sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/http/HttpConnectionReleaseDownloader.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+package com.microsoft.appcenter.distribute.download.http;
+
+import android.Manifest;
+import android.app.Notification;
+import android.app.NotificationManager;
+import android.content.Context;
+import android.net.Uri;
+import android.os.Build;
+import android.support.annotation.NonNull;
+
+import com.microsoft.appcenter.distribute.DistributeConstants;
+import com.microsoft.appcenter.distribute.PermissionUtils;
+import com.microsoft.appcenter.distribute.R;
+import com.microsoft.appcenter.distribute.ReleaseDetails;
+import com.microsoft.appcenter.distribute.download.ReleaseDownloader;
+import com.microsoft.appcenter.utils.AppCenterLog;
+import com.microsoft.appcenter.utils.AsyncTaskUtils;
+import com.microsoft.appcenter.utils.NetworkStateHelper;
+import com.microsoft.appcenter.utils.storage.SharedPreferencesManager;
+
+import java.io.File;
+import java.util.ArrayList;
+
+import static com.microsoft.appcenter.distribute.DistributeConstants.LOG_TAG;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_RELEASE_FILE;
+
+/**
+ * Downloads new releases on Android SDK versions prior to Lollipop.
+ * Uses HttpConnection and AsyncTasks.
+ */
+public class HttpConnectionReleaseDownloader implements ReleaseDownloader {
+
+    private final Context mContext;
+
+    private final ReleaseDetails mReleaseDetails;
+
+    private final ReleaseDownloader.Listener mListener;
+
+    private File mTargetFile;
+
+    private Notification.Builder mNotificationBuilder;
+
+    private HttpDownloadFileTask mHttpDownloadFileTask;
+
+    public HttpConnectionReleaseDownloader(@NonNull Context context, @NonNull ReleaseDetails releaseDetails, @NonNull ReleaseDownloader.Listener listener) {
+        mContext = context;
+        mReleaseDetails = releaseDetails;
+        mListener = listener;
+    }
+
+    private File getTargetFile() {
+        if (mTargetFile == null) {
+            mTargetFile = new File(DistributeConstants.getDownloadFilesPath(mContext), mReleaseDetails.getReleaseHash() + ".apk");
+        }
+        return mTargetFile;
+    }
+
+    private NotificationManager getNotificationManager() {
+        return (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
+    }
+
+    @NonNull
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    private Notification.Builder getNotificationBuilder() {
```

Comment:
Do not cache that.

Reply:
We must. I'll add a comment right in the code